### ISSUE TITLE
Requesting UDP leaves application in infinite loop

### ISSUE THREAD

**Issue by adventureloop** [NEAT-project/neat#125]
run

    ./server_echo -P NEAT_PROPTERY_UDP_REQUIRED

enjoy the output

**Comment by adventureloop** [comment]
Marked as help wanted, the he code is rather hairy and it might take me a while to track this down. I think I know where the loop is coming in, but I haven't seen it anywhere else.

**Comment by oystedal** [comment]
I may have worked around this in #124, `./server_echo` works fine there, even with UDP. Give it a shot.

**Comment by adventureloop** [comment]
Some debug tracing shows he is requesting creating a socket with AF: 2 Type: 1 Proto: 17.

**Comment by oystedal** [comment]
Did it work as expected in #124?

**Comment by adventureloop** [comment]
@oystedal I don't get dropped into a loop, but I can't create a UDP Connection.


[udpconnect.txt](https://github.com/NEAT-project/neat/files/385865/udpconnect.txt)

**Comment by adventureloop** [comment]
Connect fails on #124 aswell

**Comment by oystedal** [comment]
Just tested myself, same result as you when trying to connect with `./examples/client -P "NEAT_PROPERTY_UDP_REQUIRED" ...`

I'm taking a long shot, but these lines in `neat_he.c` might be to blame:

```C
            switch (stacks[i]) {
            case NEAT_STACK_UDP:
            case NEAT_STACK_UDPLITE:
                he_ctx->ai_socktype = SOCK_DGRAM;
            default:
                he_ctx->ai_socktype = SOCK_STREAM;
            }
```

**Comment by adventureloop** [comment]
Ah, adding the missing `break;` seems to resolve this.

**Comment by oystedal** [comment]
Adding the `break;` resolves the error, but I'm still not able to communicate between `./client` and `./server_echo` over UDP. I don't see those packets in wireshark either. `socat` -> `server_echo` works.

If using both `./client` and `./server_echo` over UDP works for you, then I suppose we can close this issue. Feel free to commit the missing break(s).

**Comment by adventureloop** [comment]
I ran `nc -lu 127.0.0.1 6969` and recv a messaged typed into `client`. I can respond with nc and see the message in the client output.

Using `server_echo` is not quite as good. `tcpdump` can see the client sending and the server echoing, but the client never prints the recv'd message. `server_echo` is getting errors and crashing.

I was working on test for udp when I found this regression. I will close this issue and fix problems with the udp paths as I find them when writing my test program.
